### ISSUE TITLE
Best practices document

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#24597]
The next document I'd like to work on for my [CZI grant](https://groups.google.com/g/sympy/c/vYsavewGj1w/m/nMSHqnSdBQAJ) is a document outlining some of the best practices for people using SymPy. The document will subsume this wiki page https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns, but there are also many other things that can be included. Let me know if you have any ideas for things that should be talked about. My list so far includes:

- Defining Symbols (i.e., https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#creating-symbols)
- Avoid using strings as input
- Use exact rational numbers instead of floats
- Prefer direct simplification routines to simplify(), and always avoid simplify() in library code
- For custom objects
  - Follow the args invariant (`expr.func(*expr.args) == expr`, `all(isinstance(i, Basic) for i in expr.args)`)
  - Don't denest collections (e.g., why `FiniteSet({1, 2, 3})` should produce `{{1, 2, 3}}` and not `{1, 2, 3}`)
  - Avoid over automatic simplification

The idea is to list things that apply in general. Things that are specific to a specific function should rather go in that function's documentation (although maybe that can be briefly mentioned on the page with a cross-reference). For instance, [this](https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#wild-and-match) should probably just go in the `Wild` docstring.

I also want to make a distinction between playing with SymPy interactively, where a lot of these "hacks" are OK, and code that is written down and meant to be reused.

**Comment by oscarbenjamin** [comment]
For custom objects:

1. Don't add attributes to the object. All data should be stored in args.
2. Don't add `__eq__` methods: if data is in args then this is not needed for structural equality and don't try to fight structural equality.

**Comment by sylee957** [comment]
I would also want to add how to use inequalities properly
like should we actually recommend not very pythonic tricks like 
- `if (expr1 > expr2) is S.true` instead of `if expr1 > expr2`
- `(expr1 > expr2) & (expr3 > expr4)` instead of `(expr1 > expr2) and (expr3 > expr4)`

because of the behavior of sympy inequalities?

**Comment by asmeurer** [comment]
We should also mention some general best practices for using assumptions, which may just be a cross-reference to the [assumptions guide](https://docs.sympy.org/latest/guides/assumptions.html).

**Comment by sylee957** [comment]
I think that `x = symbols('x')` is not the best practice though if we have alternative ways like `x = Symbol('x')`
and also, this unpacking behavior has to be implemented manually, and forsaking `x, = symbols('x')`.

**Comment by asmeurer** [comment]
The rationale for `symbols` is that it's a single function that works for all use-cases, so users only need to learn one thing. `Symbol` only works to create a single symbol, so to create multiple symbols at once you still have to use `symbols`.

It does seem like there's not general consensus about it among SymPy developers, however. Even the doctests still use `sympy.abc`. So it may be best to omit this from this guide at this point, or if we do include it, to mention multiple options without giving a single recommendation.

**Comment by asmeurer** [comment]
Anyone have any suggestions for good examples of things that break when you use floats instead of rationals?

**Comment by oscarbenjamin** [comment]
> Anyone have any suggestions for good examples of things that break when you use floats instead of rationals?

Are you looking for examples of things that are definitely not bugs?

Lots of polynomial related things can't work properly with floats:
```python
In [21]: e = (x + 0.1)*(x - 0.0001)

In [22]: e
Out[22]: (x - 0.0001)⋅(x + 0.1)

In [23]: e.expand()
Out[23]: 
 2                    
x  + 0.0999⋅x - 1.0e-5

In [24]: e.expand().factor()
Out[24]: 
    ⎛     2                    ⎞
1.0⋅⎝1.0⋅x  + 0.0999⋅x - 1.0e-5⎠
```

**Comment by sylee957** [comment]
I would just summarize that `Float` should be rigid for **evaluation**-type operations
(`eval, expand, simplify, rewrite`)
And it should be fragile for **solving**-type operations (`solve, dsolve, factor`)
And the problem is, things like `sum, integrate` should also be fragile for floats, however, it is hard to find examples because they sometimes work correctly because some heuristics recognize them.

**Comment by sylee957** [comment]
I also stumble across whether using symbolic integers for indexing or finite summation should be considered good practice or some antipattern to avoid.

- `summation(f(i), (i, 1, 3))` vs `sum(f(i) for i in range(1, 4))`
- `for i in range(Integer(n))`
- `list[Integer(i)], Matrix[Integer(i), Integer(j)]`

I rarely use these sort of things in production because sympy integers are a bit heavy and redundant,
but the problem is that it looks like some people can abuse sympy's sum like  https://groups.google.com/g/sympy/c/f_j0UplbtmE/m/qHAYnjxOBgAJ

**Comment by smichr** [comment]
This is probably more dev-directed, but when working on writing assumptions (as Oscar reminds frequently): don't create new expressions, only deal with sub expressions of the object.

I would like to promote use of `ordered` for finding a canonical "smallest" value or for creating a consistently ordered iterable. The idiom would be to make sure that routines are iterating over items in a canonical way if order of processing might change the result.

**Comment by smichr** [comment]
We don't have a lot of support for minimizing symbols when solving equations. I cringe when I see people give a system of symbolic expressions with lots of extraneous symbols and wonder why the solver is so slow in finding a solution. But I'm not sure if this should be added to the docs or not -- a lot of that can happen transparently, but still...it is a good practice not to load up the solver with more information than necessary.

**Comment by asmeurer** [comment]
> This is probably more dev-directed, but when working on writing assumptions (as Oscar reminds frequently): don't create new expressions, only deal with sub expressions of the object.

Yes, this is already mentioned in the assumptions and custom functions guide, but I can bring it up again here. 

> I would like to promote use of ordered for finding a canonical "smallest" value or for creating a consistently ordered iterable. The idiom would be to make sure that routines are iterating over items in a canonical way if order of processing might change the result.

I'm not sure if I completely agree with this. I think we do too much sorting in the SymPy internals. Excessive sorting has an impact on performance (especially the automatic args sorting that happens in the core). And it too often just masks bugs because the bug would only be triggered by an expression that gets ordered differently from the ones that are tested. 

With that being said, I do agree that it is nice for the same function to always return the same result on the same expression (that wouldn't necessarily happen if set ordering were used).

This does remind me of another best practice, which is to be careful about doing sorting and order comparisons on expressions, since symbolic relationals don't have an implicit ordering.

> We don't have a lot of support for minimizing symbols when solving equations. I cringe when I see people give a system of symbolic expressions with lots of extraneous symbols and wonder why the solver is so slow in finding a solution. But I'm not sure if this should be added to the docs or not -- a lot of that can happen transparently, but still...it is a good practice not to load up the solver with more information than necessary.

This seems like a hard problem, both for users to work around and to do automatically. SymPy should definitely just work no matter how big constant subexpressions are, and feels like a failing if we have to tell users otherwise. 

If it's a situation where you are calling a simplification function and only really want it to apply to the symbol(s) of interest, you should just pass those through to the function. But the problem is that you generally do require some simplification even from constants, because you need to know if they cancel out. That means that you can't just replace constant expressions with dummies, because the dummies might not cancel when the constant expressions would. I think what we need is a more targeted constant folding that replaces expressions that aren't likely to cancel with dummies, and leaves the rest intact. 

This isn't just an issue for solve. It comes up all over the place: matrices, integration, etc. It's probably worth discussing in a separate issue actually.

**Comment by asmeurer** [comment]
> This is probably more dev-directed, but when working on writing assumptions (as Oscar reminds frequently): don't create new expressions, only deal with sub expressions of the object.

Is there a good concrete example of how this can go wrong that I can give? I know it's inefficient regardless, but I'm unclear how to actually make this give an infinite recursion. Is that something that realistically only happens in the core (Add/Mul/Pow) assumptions handlers?

A related rule is that handlers should not recursively evaluate assumptions on `self`, only on the args of `self` (the global assumptions inference will handle implications between assumptions). It occurs to me that this is somewhat limiting in the old assumptions. It's impossible to state predicates on top-level facts that aren't already globally true. For example, how can you define a function that is real exactly when it is an integer (i.e., `f(x).is_real iff f(x).is_integer`)? Naively you might try

```py
class f(Function):
    def _eval_is_real(self):
        return self.is_integer
```

But this doesn't work: it leads to infinite recursion. As far as I can tell, this sort of thing is impossible to represent in the old assumptions, unless you know about which exact assumptions on `x` lead to `f(x).is_integer`. In the new assumptions, it's trivial to represent a fact like this:

```py
Equivalent(Q.real(f(x)), Q.integer(f(x)))
```

(the current support for representing quantification over `x` isn't great, but that's a separate issue)

**Comment by asmeurer** [comment]
Perhaps a simpler example is `factor(x**2.0 - 1)`. I'm not sure if we would ever want to make that work. And it showcases that floats are particularly bad in exponents.

I do remember solve used to hang if given a floating-point exponent because it would translate it into a huge degree polynomial, but I guess this has been fixed? I cannot reproduce it. I also remember there were some integrals that didn't work with floats but I can't figure one out. I did notice that manualintegrate apparently unconditionally calls `nsimplify` on its inputs.
